streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala

```diff
@@ -821,6 +821,33 @@ class CheckpointSuite extends TestSuiteBase with DStreamCheckpointTester
     checkpointWriter.stop()
   }
 
+  test("SPARK-6847: stack overflow when updateStateByKey is followed by a checkpointed dstream") {
+    ssc = new StreamingContext(master, framework, batchDuration)
+    val batchCounter = new BatchCounter(ssc)
+    ssc.checkpoint(checkpointDir)
+    val inputDStream = new CheckpointInputDStream(ssc)
+    val updateFunc = (values: Seq[Int], state: Option[Int]) => {
+      Some(values.sum + state.getOrElse(0))
+    }
+    @volatile var recursiveCheckpoint = false
+    @volatile var rddsBothCheckpointed = false
+    inputDStream.map(i => (i, i)).
+      updateStateByKey[Int](updateFunc).checkpoint(batchDuration).
+      map(i => i).checkpoint(batchDuration).
```

Comment:
would you test this using two updateStateByKey

also this code style is weird and hard to read.